File: Java_Solutions/Word_Search_II.java
/*
* Leetcode Challenge: Word Search II
*
* Given an m x n board of characters and a list of strings words, 
* return all words on the board.
*
* Each word must be constructed from letters of sequentially adjacent cells,
* where adjacent cells are horizontally or vertically neighboring. 
* The same letter cell may not be used more than once in a word.
*
* 9/8/24
* Mark Robuck
*/
class TrieNode {
    TrieNode[] children = new TrieNode[26];
    String word = null;
}


public class Solution {
    /*
    * Find all words in a list given a word search board.
    * @params board, given word search board
    * @params words, list of strings representing valid words
    * @return list of all strings found in board from words
    */
    public List<String> findWords(char[][] board, String[] words) {
        //build a tree from word list, words
        TrieNode root = new TrieNode();
        for (String word : words) {
            TrieNode node = root;
            for (char c : word.toCharArray()) {
                int index = c - 'a';
                if (node.children[index] == null) {
                    node.children[index] = new TrieNode();
                }
                node = node.children[index];
            }
            node.word = word;
        }

        //perform depth first search on board
        List<String> result = new ArrayList<>();
        for (int i = 0; i < board.length; i++) {
            for (int j = 0; j < board[0].length; j++) {
                dfs(board, i, j, root, result);
            }
        }
        
        return result;
    }

        //depth first search
        private void dfs(char[][] board, int i, int j, TrieNode node, List<String> result) {
            char c = board[i][j];
            if (c == '#' || node.children[c - 'a'] == null) return;
            
            node = node.children[c - 'a'];
            if (node.word != null) {
                result.add(node.word);
                //avoid duplicate words
                node.word = null;
            }

        //mark cells as visited
        board[i][j] = '#';
        if (i > 0) dfs(board, i - 1, j, node, result);
        if (i < board.length - 1) dfs(board, i + 1, j, node, result);
        if (j > 0) dfs(board, i, j - 1, node, result);
        if (j < board[0].length - 1) dfs(board, i, j + 1, node, result);
        //reset the cell after search
        board[i][j] = c;  
    }
}

File: Python_Solutions/Word_Search_II.py
"""
Leetcode Challenge: Word Search II

Given an m x n board of characters and a list of strings words, 
return all words on the board.

Each word must be constructed from letters of sequentially adjacent cells,
 where adjacent cells are horizontally or vertically neighboring. 
 The same letter cell may not be used more than once in a word.

9/8/24
Mark Robuck
"""
class TrieNode:
    def __init__(self):
        self.children = {}
        self.word = None


class Solution:
    """
    Find all words in a list given a word search board.
    Args:
        board(List[str]): given word search board
        words(List[str]): list of strings representing valid words
    Returns:
        List[str]: list of all strings found in board from words
    """
    def findWords(self, board: List[List[str]], words: List[str]) -> List[str]:
        #build a tree from word list, words
        root = TrieNode()
        for word in words:
            node = root
            for char in word:
                if char not in node.children:
                    node.children[char] = TrieNode()
                node = node.children[char]
            node.word = word

        #depth first search
        def dfs(board, i, j, node, res):
            char = board[i][j]
            if char not in node.children:
                return
            node = node.children[char]
            if node.word:
                res.append(node.word)
                #avoid duplicate words
                node.word = None

            #mark cells as visited
            board[i][j] = '#'
            for x, y in [(i-1, j), (i+1, j), (i, j-1), (i, j+1)]:
                if 0 <= x < len(board) and 0 <= y < len(board[0]) and board[x][y] != '#':
                    dfs(board, x, y, node, res)
            #reset the cell after search
            board[i][j] = char

        #perform DFS on board
        res = []
        for i in range(len(board)):
            for j in range(len(board[0])):
                dfs(board, i, j, root, res)
        
        return res
        
